File: np_platform_message_relay/ios/Classes/MessageRelayChannelHandler.swift
import Flutter
import Foundation
import NpIosCore

class MessageRelayChannelHandler: NSObject, FlutterStreamHandler {
  override init() {
    do {
      Self.idLock.lock()
      defer { Self.idLock.unlock() }
      id = Self.nextId
      Self.nextId += 1
    }
    super.init()
  }

  func onMethodCall(call: FlutterMethodCall, result: FlutterResult) {
    let args = call.arguments as? Dictionary<String, Any>
    do {
      switch call.method {
      case "broadcast":
        try broadcast(
          event: (args?["event"] as? String).unwrap(),
          data: args?["data"] as? String,
          result: result
        )

      default:
        result(FlutterMethodNotImplemented)
      }
    } catch let error as AppError {
      result(FlutterError(code: "systemException", message: error.message, details: "\(error.stackTrace)"))
    } catch {
      result(FlutterError(code: "systemException", message: "\(error)", details: nil))
    }
  }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    Self.evLock.lock()
    defer { Self.evLock.unlock() }
    Self.eventSinks[id] = events
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    Self.evLock.lock()
    defer { Self.evLock.unlock() }
    Self.eventSinks.removeValue(forKey: id)
    return nil
  }

  private func broadcast(event: String, data: String?, result: FlutterResult) {
    Self.evLock.lock()
    defer { Self.evLock.unlock() }
    for s in Self.eventSinks.values {
      var map = ["event": event]
      if data != nil {
        map["data"] = data
      }
      s(data)
    }
    result(nil)
  }

  static let eventChannel = "\(K.libId)/message_relay_event"
  static let methodChannel = "\(K.libId)/message_relay_method"

  private static var eventSinks: [Int:FlutterEventSink] = [:]
  private static var evLock = NSRecursiveLock()
  private static var nextId = 0
  private static var idLock = NSRecursiveLock()

  private let id: Int
}

File: np_platform_message_relay/android/src/main/kotlin/com/nkming/nc_photos/np_platform_message_relay/MessageRelayChannelHandler.kt
package com.nkming.nc_photos.np_platform_message_relay

import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

internal class MessageRelayChannelHandler : MethodChannel.MethodCallHandler,
	EventChannel.StreamHandler {
	companion object {
		const val EVENT_CHANNEL = "${K.LIB_ID}/message_relay_event"
		const val METHOD_CHANNEL = "${K.LIB_ID}/message_relay_method"

		private val eventSinks = mutableMapOf<Int, EventChannel.EventSink>()
		private var nextId = 0
	}

	override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
		when (call.method) {
			"broadcast" -> {
				try {
					broadcast(
						call.argument("event")!!, call.argument("data"), result
					)
				} catch (e: Throwable) {
					result.error("systemException", e.toString(), null)
				}
			}

			else -> {
				result.notImplemented()
			}
		}
	}

	override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
		synchronized(eventSinks) {
			eventSinks[id] = events
		}
	}

	override fun onCancel(arguments: Any?) {
		synchronized(eventSinks) {
			eventSinks.remove(id)
		}
	}

	private fun broadcast(
		event: String, data: String?, result: MethodChannel.Result
	) {
		synchronized(eventSinks) {
			for (s in eventSinks.values) {
				s.success(buildMap {
					put("event", event)
					if (data != null) put("data", data)
				})
			}
		}
		result.success(null)
	}

	private val id = nextId++
}
